### PR TITLE
Fixed error with batchFirst assertion

### PR DESCRIPTION
**Pull request by SeanNaren**
Hey,

Found an error when calling updateGradInput twice with batchFirst, problem can be replicated using this:
```
require 'cudnn'
blstm = cudnn.BLSTM(2,5,1,true)
input = torch.rand(6,3,2):cuda()
output = blstm:forward(input)
grad = blstm:updateGradInput(input,output)
grad = blstm:updateGradInput(input,output)
```

### CODE REVIEW

**Comment by SeanNaren** [comment]
Hmm I'm starting to think that this might actually not be an issue. Should a user be allowed to called GradInput twice after forwarding once through an RNN?

**Comment by ngimel** [comment]
There problems with batchFirst regardless, e.g. lines 238-240 
```
if (self.batchFirst) then
         input = input:transpose(1, 2)
end
```
won't work. You are not transposing your underlying data here, you are just changing strides and sizes of the tensor, so you'll get wrong results. Proper transposition would involve creating (or reusing) new tensor, something like self.input_, and copying transposed data into it:
```
self.input_:input(input:transpose(1,2))
```

**Comment by ngimel** [comment]
Ignore my previous comment, makeContiguous is later called on input, doing just that. Sorry.

**Comment by SeanNaren** [comment]
@ngimel No worries if anything helped explain how the stuff works!

Do you feel like the point of the PR is important to address, being able to call gradOutput twice without a forward call?

**Comment by soumith** [comment]
@SeanNaren yes it is. We do gradient descent on the input in a few use-cases, like finding what a particular neuron is most sensitive to.

**Comment by soumith** [comment]
and in those cases, multiple gradInput calls are issued.

**Comment by SeanNaren** [comment]
@soumith great, just give me some time before pulling this in, I think I can do this in a better way that doesn't involve messing with the output itself.

**Comment by ngimel** [comment]
BTW, makeContiguous is called on input in updateGradOutput and accGradParameters, which for batchFirst would result in an unnecessary copy/transpose of input.

**Comment by SeanNaren** [comment]
@ngimel not entirely sure what you mean, do we not need to transpose the input before proceeding?

**Comment by ngimel** [comment]
If you look at the code, x (contiguous input) is not used in updateGradOutput and accGradParameters. input:transpose(1,2) in the beginning is necessary (cheap operation), makeContiguous(input) is not (data copy, expensive operation).

**Comment by SeanNaren** [comment]
@ngimel Hows it looking now? I think x is used in accGradParams though, let me know if it does need changing though. Thanks for your help!

File: RNN.lua [hunk 1]
```diff
@@ -450,7 +448,12 @@ function RNN:accGradParameters(input, gradOutput, scale)
    assert(input:size(2) == self.miniBatch, 'input has incorrect minibatch size!')
    assert(input:size(3) == self.inputSize, 'input has incorrect size!')
 
-   assert(gradOutput:isSameSizeAs(self.output), 'gradOutput has incorrect size!')
+    if (self.batchFirst) then
```
Comment (ngimel) [inline RNN.lua]: Don't you want to make lines 451-456 similar to lines 369-370? Otherwise looks good, you are right, x is used in accGradParameters.

**Comment by SeanNaren** [comment]
Thanks nice catch, should hopefully be everything. Thanks for your help again

**Comment by ngimel** [comment]
@soumith non-contiguous output and gradInput are returned if batchFirst==true. Is it Ok, or should they be made contiguous?

**Comment by soumith** [comment]
@ngimel i think it's fine, i dont see why they should be made contiguous.

**Comment by SeanNaren** [comment]
Awesome, hopefully that should fix up batchFirst! I've added another feature to extract biases and weights for each layer and each linear layer. Just need to verify its correctness and maybe update the tests.

@ngimel could you have a look at the added functionality? Had some requests for being able to extract the params. Not sure where I should add this, but I would definitely like to add some documentation on how to read the outputs of these methods (they basically do the same as the tests, except put the tensors into separate tables). Thanks!

**Comment by ngimel** [comment]
@SeanNaren, looks good but let's hold off with this until cudnn v5 final is available with its API changes. Then in retrieveLinearParams xDescs will have to be replaced by xDescs[0].

**Comment by SeanNaren** [comment]
Sounds good, just reverted the added stuff, everything else should be good to go!

**Comment by SeanNaren** [comment]
@soumith hows this looking? Let me know if you have any feedback on this!

**Comment by soumith** [comment]
@SeanNaren does this work with cudnn R5 final? Can you rebase your PR onto the top of the R5 head and re-test?

**Comment by soumith** [comment]
the bindings in the R5 branch have been updated for R5 final.

**Comment by SeanNaren** [comment]
@soumith Can confirm it works, the rebase version can be seen [here](https://github.com/SeanNaren/cudnn.torch/tree/R5Rebase)

**Pull request merged**